File: scripts/publish-global-office.py
#!/usr/bin/env python3
"""
publish-global-office.py
────────────────────────
Picks the 2 oldest unpublished Global Office markdown articles from the
project folder and posts them to Sanity CMS as published articles.

Tracking: scripts/.published-global-office.json  (list of filenames already sent)

Usage:
  python3 scripts/publish-global-office.py            # publish 2 articles
  python3 scripts/publish-global-office.py --dry-run  # preview without posting
  python3 scripts/publish-global-office.py --count 1  # publish N articles
"""

import argparse
import json
import os
import re
import sys
import uuid
from datetime import datetime, timezone
from pathlib import Path

import requests
import yaml

# ── Paths & config ────────────────────────────────────────────────────────────

ROOT = Path(__file__).resolve().parent.parent   # the-boardroom-brief/
TRACKING = Path(__file__).resolve().parent / ".published-global-office.json"

SANITY_PROJECT_ID = os.environ.get("NEXT_PUBLIC_SANITY_PROJECT_ID", "e8dwtkci")
SANITY_DATASET    = os.environ.get("NEXT_PUBLIC_SANITY_DATASET", "production")
SANITY_TOKEN      = os.environ.get("SANITY_API_TOKEN")
SITE_URL          = os.environ.get("NEXT_PUBLIC_SITE_URL", "https://thealignmenttimes.com")

SANITY_MUTATE = (
    f"https://{SANITY_PROJECT_ID}.api.sanity.io"
    f"/v2024-01-01/data/mutate/{SANITY_DATASET}"
)
SANITY_QUERY = (
    f"https://{SANITY_PROJECT_ID}.api.sanity.io"
    f"/v2024-01-01/data/query/{SANITY_DATASET}"
)

PILLAR_ID   = "global-office"
PILLAR_NAME = "The Global Office"
AUTHOR_ID   = "author-priya-mehta"
AUTHOR_NAME = "Priya Mehta"


# ── Markdown parser ───────────────────────────────────────────────────────────

def key() -> str:
    return uuid.uuid4().hex[:12]


def parse_inline(text: str) -> list[dict]:
    """
    Convert inline markdown (bold, italic) to Sanity span children.
    Returns a list of span dicts.
    """
    spans = []
    # Pattern: **bold**, *italic*, or plain text
    pattern = re.compile(r"\*\*(.+?)\*\*|\*(.+?)\*|([^*]+)")
    for m in pattern.finditer(text):
        if m.group(1):  # **bold**
            spans.append({"_type": "span", "_key": key(), "text": m.group(1), "marks": ["strong"]})
        elif m.group(2):  # *italic*
            spans.append({"_type": "span", "_key": key(), "text": m.group(2), "marks": ["em"]})
        else:            # plain
            t = m.group(3)
            if t:
                spans.append({"_type": "span", "_key": key(), "text": t, "marks": []})
    return spans or [{"_type": "span", "_key": key(), "text": text, "marks": []}]


def markdown_to_blocks(body: str) -> list[dict]:
    """
    Convert markdown body to Sanity portable text blocks.
    - ## → h2,  ### → h3
    - ![alt](url) → skipped (hero/body images handled via coverImage field)
    - *italic / **bold** inline marks supported
    - Byline lines (*By …*) and caption lines (*Illustration:…*) → italic block
    - Blank lines separate blocks
    """
    blocks = []
    raw_paras = re.split(r"\n{2,}", body.strip())

    for para in raw_paras:
        para = para.strip()
        if not para:
            continue

        # Skip image lines entirely — stored as coverImage / ogImage
        if re.match(r"^!\[", para):
            continue

        # Heading ## or ###
        h3 = re.match(r"^###\s+(.+)", para)
        h2 = re.match(r"^##\s+(.+)", para)
        if h3:
            blocks.append({
                "_type": "block", "_key": key(), "style": "h3",
                "markDefs": [],
                "children": parse_inline(h3.group(1)),
            })
            continue
        if h2:
            blocks.append({
                "_type": "block", "_key": key(), "style": "h2",
                "markDefs": [],
                "children": parse_inline(h2.group(1)),
            })
            continue

        # Multi-line paragraph — join lines
        lines = para.splitlines()
        merged = " ".join(l.strip() for l in lines if l.strip())

        # Skip image-only lines that slipped through
        if re.match(r"^!\[", merged):
            continue

        blocks.append({
            "_type": "block", "_key": key(), "style": "normal",
            "markDefs": [],
            "children": parse_inline(merged),
        })

    return blocks


# ── Frontmatter parser ────────────────────────────────────────────────────────

def parse_article(path: Path) -> dict | None:
    """Parse a markdown file into a structured dict. Returns None on failure."""
    text = path.read_text(encoding="utf-8")
    fm_match = re.match(r"^---\n(.*?)\n---\n(.*)", text, re.DOTALL)
    if not fm_match:
        print(f"  ⚠ No frontmatter found in {path.name} — skipping")
        return None

    try:
        fm = yaml.safe_load(fm_match.group(1))
    except yaml.YAMLError as e:
        print(f"  ⚠ YAML error in {path.name}: {e} — skipping")
        return None

    body_md = fm_match.group(2).strip()

    # Extract title from first H1 line (# Title) or first ## heading
    title_match = re.search(r"^#\s+(.+)$", body_md, re.MULTILINE)
    title = title_match.group(1).strip() if title_match else path.stem.replace("-", " ").title()

    # Satirical headline: the italic byline "By Priya Mehta, The Global Office"
    # → extract first paragraph-level italicised line after the H1 as excerpt instead
    excerpt_match = re.search(
        r"^(?:#{1,3}.+\n+)?(.+?)(?:\n\n|\Z)", body_md.lstrip("# \n"), re.DOTALL
    )
    excerpt = ""
    if excerpt_match:
        raw = excerpt_match.group(1).strip()
        # Strip markdown formatting for plain excerpt
        raw = re.sub(r"\*\*(.+?)\*\*", r"\1", raw)
        raw = re.sub(r"\*(.+?)\*", r"\1", raw)
        raw = re.sub(r"!\[.*?\]\(.*?\)", "", raw)
        excerpt = raw[:300].strip()

    # Hero image URL from frontmatter
    images = fm.get("images", {}) or {}
    hero_url = images.get("hero", "")
    og_url   = images.get("body", hero_url)

    # Countries list
    countries_raw = fm.get("countries", "")
    if isinstance(countries_raw, str):
        countries = [c.strip() for c in countries_raw.split(",") if c.strip()]
    elif isinstance(countries_raw, list):
        countries = countries_raw
    else:
        countries = []

    # Published date
    date_val = fm.get("date")
    if isinstance(date_val, str):
        pub_dt = datetime.fromisoformat(date_val).replace(
            hour=8, tzinfo=timezone.utc
        ).isoformat()
    elif date_val:
        pub_dt = datetime(date_val.year, date_val.month, date_val.day,
                          8, 0, 0, tzinfo=timezone.utc).isoformat()
    else:
        pub_dt = datetime.now(timezone.utc).isoformat()

    # Portable text body (strip the H1 title line to avoid duplication)
    body_no_title = re.sub(r"^#\s+.+\n*", "", body_md, count=1).strip()
    blocks = markdown_to_blocks(body_no_title)

    # Estimate read time
    word_count = fm.get("word_count") or len(body_md.split())
    read_time = max(1, round(int(word_count) / 200))

    # Slug
    slug_base = re.sub(r"[^a-z0-9]+", "-", title.lower()).strip("-")[:96]

    # Sources for seoDescription
    subject = fm.get("subject", "")
    c_str   = " vs ".join(countries[:2]) if len(countries) >= 2 else ", ".join(countries)
    seo_desc = f"{c_str}: {subject}. Analysis by Priya Mehta for The Alignment Times."[:160]

    return {
        "filename":    path.name,
        "title":       title,
        "slug":        slug_base,
        "excerpt":     excerpt,
        "subject":     subject,
        "countries":   countries,
        "hero_url":    hero_url,
        "og_url":      og_url,
        "pub_dt":      pub_dt,
        "read_time":   read_time,
        "blocks":      blocks,
        "seo_desc":    seo_desc,
        "word_count":  word_count,
    }


# ── Sanity helpers ────────────────────────────────────────────────────────────

def sanity_headers() -> dict:
    return {
        "Authorization": f"Bearer {SANITY_TOKEN}",
        "Content-Type":  "application/json",
    }


def ensure_pillar() -> None:
    """Create the global-office pillar doc if it doesn't exist."""
    mutations = [{
        "createIfNotExists": {
            "_id":   PILLAR_ID,
            "_type": "pillar",
            "name":  PILLAR_NAME,
            "slug":  {"_type": "slug", "current": PILLAR_ID},
        }
    }]
    r = requests.post(SANITY_MUTATE, headers=sanity_headers(),
                      json={"mutations": mutations}, timeout=15)
    r.raise_for_status()


def ensure_author() -> None:
    """Create the Priya Mehta author doc if it doesn't exist."""
    mutations = [{
        "createIfNotExists": {
            "_id":   AUTHOR_ID,
            "_type": "author",
            "name":  AUTHOR_NAME,
            "slug":  {"_type": "slug", "current": "priya-mehta"},
            "bio":   "Staff writer for The Alignment Times covering global work and life culture.",
        }
    }]
    r = requests.post(SANITY_MUTATE, headers=sanity_headers(),
                      json={"mutations": mutations}, timeout=15)
    r.raise_for_status()


def publish_article(article: dict, dry_run: bool = False) -> str:
    """
    Upsert the article to Sanity. Returns the published URL.
    Uses createOrReplace so re-running is idempotent.
    """
    doc_id = f"article-{article['slug']}"

    # Country reference IDs: use slugified name as stable _id
    country_refs = []
    for c in article["countries"]:
        cid = "country-" + re.sub(r"[^a-z0-9]+", "-", c.lower()).strip("-")
        country_refs.append({"_type": "reference", "_ref": cid, "_key": key()})

    doc = {
        "_id":              doc_id,
        "_type":            "article",
        "title":            article["title"],
        "slug":             {"_type": "slug", "current": article["slug"]},
        "excerpt":          article["excerpt"],
        "body":             article["blocks"],
        "pillar":           {"_type": "reference", "_ref": PILLAR_ID},
        "author":           {"_type": "reference", "_ref": AUTHOR_ID},
        "countries":        country_refs,
        "publishedAt":      article["pub_dt"],
        "readTime":         article["read_time"],
        "featured":         False,
        "aiGenerated":      True,
        "agentName":        "Priya Mehta",
        # Image fields (URL-based, not Sanity asset uploads)
        "ogImage":          article["og_url"] or article["hero_url"],
        "imageGeneratedWith": "pexels",
        # SEO
        "seoDescription":   article["seo_desc"],
    }

    published_url = f"{SITE_URL}/global-office/{article['slug']}"

    if dry_run:
        print(f"  [dry-run] Would create: {doc_id}")
        print(f"            Title:  {article['title']}")
        print(f"            Slug:   {article['slug']}")
        print(f"            Date:   {article['pub_dt']}")
        print(f"            Blocks: {len(article['blocks'])}")
        print(f"            URL:    {published_url}")
        return published_url

    mutations = [{"createOrReplace": doc}]
    r = requests.post(
        SANITY_MUTATE,
        headers=sanity_headers(),
        json={"mutations": mutations},
        timeout=20,
    )
    r.raise_for_status()
    result = r.json()
    if result.get("error"):
        raise RuntimeError(f"Sanity error: {result['error']}")

    # Trigger ISR revalidation (non-fatal)
    try:
        revalidate_secret = os.environ.get("REVALIDATE_SECRET", "")
        requests.post(
            f"{SITE_URL}/api/revalidate?secret={revalidate_secret}&path=/global-office",
            timeout=5,
        )
    except Exception:
        pass

    return published_url


# ── Country doc upserts ───────────────────────────────────────────────────────

def ensure_countries(countries: list[str], dry_run: bool) -> None:
    if not countries or dry_run:
        return
    mutations = []
    for c in countries:
        cid = "country-" + re.sub(r"[^a-z0-9]+", "-", c.lower()).strip("-")
        mutations.append({
            "createIfNotExists": {
                "_id":   cid,
                "_type": "country",
                "name":  c,
                "slug":  {"_type": "slug", "current": cid.replace("country-", "")},
            }
        })
    if mutations:
        r = requests.post(SANITY_MUTATE, headers=sanity_headers(),
                          json={"mutations": mutations}, timeout=15)
        r.raise_for_status()


# ── Tracking ──────────────────────────────────────────────────────────────────

def load_tracking() -> set[str]:
    if TRACKING.exists():
        return set(json.loads(TRACKING.read_text()))
    return set()


def save_tracking(published: set[str]) -> None:
    TRACKING.write_text(json.dumps(sorted(published), indent=2))


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Publish Global Office articles to Sanity")
    parser.add_argument("--dry-run", action="store_true", help="Preview without posting")
    parser.add_argument("--count", type=int, default=2, help="Number of articles to publish (default 2)")
    args = parser.parse_args()

    if not SANITY_TOKEN and not args.dry_run:
        sys.exit("❌ SANITY_API_TOKEN is not set. Export it or use --dry-run.")

    # Find all Global Office markdown files, sorted oldest-first
    md_files = sorted(
        [f for f in ROOT.glob("*.md") if f.name != "SKILL.md"],
        key=lambda f: f.name,
    )

    published = load_tracking()
    candidates = [f for f in md_files if f.name not in published]

    if not candidates:
        print("✅ All articles already published — nothing to do.")
        return

    to_publish = candidates[: args.count]
    print(f"📋 Found {len(candidates)} unpublished articles. Publishing {len(to_publish)}.\n")

    if not args.dry_run:
        print("🔧 Ensuring pillar and author exist in Sanity...")
        ensure_pillar()
        ensure_author()

    newly_published = []
    for f in to_publish:
        print(f"📄 Processing: {f.name}")
        article = parse_article(f)
        if not article:
            print(f"  ⚠ Skipped.\n")
            continue

        if not args.dry_run:
            ensure_countries(article["countries"], dry_run=False)

        url = publish_article(article, dry_run=args.dry_run)

        if not args.dry_run:
            published.add(f.name)
            newly_published.append(f.name)

        print(f"  ✓ {'Would post' if args.dry_run else 'Published'}: {article['title']}")
        print(f"    → {url}\n")

    if not args.dry_run and newly_published:
        save_tracking(published)
        print(f"✅ Done. {len(newly_published)} article(s) published. Tracking file updated.")
    elif args.dry_run:
        print(f"✅ Dry run complete. {len(to_publish)} article(s) would be published.")

    remaining = len(candidates) - len(to_publish)
    if remaining > 0:
        print(f"📬 {remaining} article(s) remain in the queue for future runs.")


if __name__ == "__main__":
    main()

File: scripts/publish-global-office.mjs
#!/usr/bin/env node
/**
 * publish-global-office.mjs
 * ─────────────────────────
 * Manual one-shot publisher for Global Office .md articles (local dev use).
 * Production publishing is handled by the cron route:
 *   app/api/cron/publish-articles/route.ts  (runs daily at 09:00 UTC)
 *
 * WARNING: The markdown parsing helpers (parseFrontmatter, markdownToBlocks,
 * parseInline, slugify) are duplicated from route.ts. If you fix a bug in
 * one, apply the same fix in the other.
 *
 * Usage (from project root):
 *   node scripts/publish-global-office.mjs            → publish 2 articles
 *   node scripts/publish-global-office.mjs --dry-run  → preview only
 *   node scripts/publish-global-office.mjs --count 1  → publish N articles
 */

import { createClient } from "@sanity/client";
import { readFileSync, writeFileSync, existsSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { readdirSync } from "fs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT      = resolve(__dirname, "..");
const TRACKING  = resolve(__dirname, ".published-global-office.json");
const DOTENV    = resolve(ROOT, ".env.local");

// ── Load .env.local ──────────────────────────────────────────────────────────

function loadEnv(path) {
  if (!existsSync(path)) return;
  const lines = readFileSync(path, "utf8").split("\n");
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq < 1) continue;
    const k = trimmed.slice(0, eq).trim();
    const v = trimmed.slice(eq + 1).trim();
    if (!process.env[k]) process.env[k] = v;
  }
}

loadEnv(DOTENV);

// ── Config ───────────────────────────────────────────────────────────────────

const PROJECT_ID = process.env.NEXT_PUBLIC_SANITY_PROJECT_ID ?? "e8dwtkci";
const DATASET    = process.env.NEXT_PUBLIC_SANITY_DATASET    ?? "production";
const TOKEN      = process.env.SANITY_API_TOKEN;
const SITE_URL   = process.env.NEXT_PUBLIC_SITE_URL ?? "https://thealignmenttimes.com";
const REVALIDATE = process.env.REVALIDATE_SECRET ?? "";

const PILLAR_ID   = "global-office";
const PILLAR_NAME = "The Global Office";
const AUTHOR_ID   = "author-priya-mehta";
const AUTHOR_NAME = "Priya Mehta";

// ── Args ─────────────────────────────────────────────────────────────────────

const args    = process.argv.slice(2);
const DRY_RUN = args.includes("--dry-run");
const COUNT   = (() => {
  const i = args.indexOf("--count");
  return i >= 0 && args[i + 1] ? parseInt(args[i + 1], 10) : 2;
})();

// ── Sanity client ─────────────────────────────────────────────────────────────

const writeClient = createClient({
  projectId: PROJECT_ID,
  dataset:   DATASET,
  apiVersion: "2024-01-01",
  useCdn:    false,
  token:     TOKEN,
});

// ── Helpers ──────────────────────────────────────────────────────────────────

let _keyCounter = 0;
function key() { return `k${++_keyCounter}`; }

function slugify(text) {
  return text.toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-|-$/g, "").slice(0, 96);
}

/** Parse inline markdown bold/italic into Sanity span children */
function parseInline(text) {
  const spans = [];
  const re = /\*\*(.+?)\*\*|\*(.+?)\*|([^*]+)/g;
  let m;
  while ((m = re.exec(text)) !== null) {
    if (m[1]) spans.push({ _type: "span", _key: key(), text: m[1], marks: ["strong"] });
    else if (m[2]) spans.push({ _type: "span", _key: key(), text: m[2], marks: ["em"] });
    else if (m[3]) spans.push({ _type: "span", _key: key(), text: m[3], marks: [] });
  }
  return spans.length ? spans : [{ _type: "span", _key: key(), text, marks: [] }];
}

/** Convert markdown body to Sanity portable text blocks */
function markdownToBlocks(body) {
  const blocks = [];
  const paras  = body.split(/\n{2,}/);

  for (let para of paras) {
    para = para.trim();
    if (!para || /^!\[/.test(para)) continue;

    // ── Headings ──────────────────────────────────────────────────────────────
    const h4 = para.match(/^####\s+(.+)/);
    const h3 = para.match(/^###\s+(.+)/);
    const h2 = para.match(/^##\s+(.+)/);
    if (h4) { blocks.push({ _type: "block", _key: key(), style: "h3", markDefs: [], children: parseInline(h4[1]) }); continue; }
    if (h3) { blocks.push({ _type: "block", _key: key(), style: "h3", markDefs: [], children: parseInline(h3[1]) }); continue; }
    if (h2) { blocks.push({ _type: "block", _key: key(), style: "h2", markDefs: [], children: parseInline(h2[1]) }); continue; }

    // ── Markdown table ────────────────────────────────────────────────────────
    // Detected when multiple lines contain | characters
    const lines = para.split("\n");
    const tableLines = lines.filter(l => l.includes("|"));
    if (tableLines.length > 1) {
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        // Skip separator rows like |---|---|
        if (/^\|[-|\s:]+\|$/.test(trimmed)) continue;
        // Extract cells, strip empty first/last from leading/trailing pipes
        const cells = trimmed.split("|").map(c => c.trim()).filter(Boolean);
        if (cells.length === 0) continue;
        const text = cells.join("  |  ");
        if (text.trim()) {
          blocks.push({ _type: "block", _key: key(), style: "normal", markDefs: [], children: parseInline(text) });
        }
      }
      continue;
    }

    // ── Normal paragraph ──────────────────────────────────────────────────────
    const merged = para.split("\n").map(l => l.trim()).filter(Boolean).join(" ");
    if (!merged || /^!\[/.test(merged)) continue;
    blocks.push({ _type: "block", _key: key(), style: "normal", markDefs: [], children: parseInline(merged) });
  }
  return blocks;
}

/** Parse YAML-ish frontmatter (handles simple key: value and nested maps) */
function parseFrontmatter(raw) {
  const fm = {};
  const lines = raw.split("\n");
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    // Nested block (e.g. images:)
    const nested = line.match(/^(\w+):\s*$/);
    if (nested) {
      const block = {};
      i++;
      while (i < lines.length && /^\s{2}/.test(lines[i])) {
        const kv = lines[i].trim().match(/^(\w+):\s*(.*)/);
        if (kv) block[kv[1]] = kv[2].trim();
        i++;
      }
      fm[nested[1]] = block;
      continue;
    }
    const kv = line.match(/^(\w[\w_]*):\s*(.*)/);
    if (kv) fm[kv[1]] = kv[2].trim();
    i++;
  }
  return fm;
}

/** Parse one .md file into a structured article object */
function parseArticle(filePath) {
  const text  = readFileSync(filePath, "utf8");
  const match = text.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) { console.warn(`  ⚠ No frontmatter: ${filePath}`); return null; }

  const fm     = parseFrontmatter(match[1]);
  const bodyMd = match[2].trim();

  // Title from first H1, fallback to filename
  const titleMatch = bodyMd.match(/^#\s+(.+)$/m);
  const title      = titleMatch ? titleMatch[1].trim() : filePath.split("/").pop().replace(/-/g," ");

  // Excerpt: first non-heading paragraph, stripped of markdown
  const bodyNoTitle = bodyMd.replace(/^#\s+.+\n*/m, "").trim();
  const firstBlock  = bodyNoTitle.split(/\n{2,}/)[0] || "";
  const excerpt     = firstBlock
    .replace(/\*\*(.+?)\*\*/g, "$1")
    .replace(/\*(.+?)\*/g, "$1")
    .replace(/!\[.*?\]\(.*?\)/g, "")
    .replace(/^#+\s+/, "")
    .trim()
    .slice(0, 300);

  // Image URLs
  const images  = fm.images || {};
  const heroUrl = images.hero || "";
  const ogUrl   = images.body || heroUrl;

  // Countries
  const countriesRaw = fm.countries || "";
  const countries    = typeof countriesRaw === "string"
    ? countriesRaw.split(",").map(c => c.trim()).filter(Boolean)
    : [];

  // Published date at 8am UTC
  const dateStr = fm.date || new Date().toISOString().split("T")[0];
  const pubDt   = new Date(`${dateStr}T08:00:00Z`).toISOString();

  // Read time
  const wordCount = parseInt(fm.word_count) || bodyMd.split(/\s+/).length;
  const readTime  = Math.max(1, Math.round(wordCount / 200));

  // Slug from title
  const slug = slugify(title);

  // Portable text
  const blocks = markdownToBlocks(bodyNoTitle);

  // SEO
  const subject = fm.subject || "";
  const cStr    = countries.slice(0, 2).join(" vs ");
  const seoDesc = `${cStr}: ${subject}. Analysis by Priya Mehta for The Alignment Times.`.slice(0, 160);

  return { title, slug, excerpt, subject, countries, heroUrl, ogUrl, pubDt, readTime, blocks, seoDesc, wordCount };
}

// ── Sanity mutations ──────────────────────────────────────────────────────────

async function ensurePillar() {
  await writeClient.createIfNotExists({
    _id: PILLAR_ID, _type: "pillar",
    name: PILLAR_NAME,
    slug: { _type: "slug", current: PILLAR_ID },
  });
}

async function ensureAuthor() {
  await writeClient.createIfNotExists({
    _id: AUTHOR_ID, _type: "author",
    name: AUTHOR_NAME,
    slug: { _type: "slug", current: "priya-mehta" },
    bio:  "Staff writer for The Alignment Times covering global work and life culture.",
  });
}

async function ensureCountries(countries) {
  for (const c of countries) {
    const cid = "country-" + slugify(c);
    await writeClient.createIfNotExists({
      _id: cid, _type: "country",
      name: c,
      slug: { _type: "slug", current: cid.replace("country-", "") },
    });
  }
}

async function publishArticle(article) {
  const docId = `article-${article.slug}`;

  const countryRefs = article.countries.map(c => ({
    _type: "reference", _ref: "country-" + slugify(c), _key: key(),
  }));

  const doc = {
    _id:              docId,
    _type:            "article",
    title:            article.title,
    slug:             { _type: "slug", current: article.slug },
    excerpt:          article.excerpt,
    body:             article.blocks,
    pillar:           { _type: "reference", _ref: PILLAR_ID },
    author:           { _type: "reference", _ref: AUTHOR_ID },
    countries:        countryRefs,
    publishedAt:      article.pubDt,
    readTime:         article.readTime,
    featured:         false,
    aiGenerated:      true,
    agentName:        "Priya Mehta",
    ogImage:          article.ogUrl || article.heroUrl,
    imageGeneratedWith: "pexels",
    seoDescription:   article.seoDesc,
  };

  await writeClient.createOrReplace(doc);

  // Trigger ISR revalidation (non-fatal)
  try {
    await fetch(`${SITE_URL}/api/revalidate?secret=${REVALIDATE}&path=/global-office`, { method: "POST" });
  } catch {}

  return `${SITE_URL}/global-office/${article.slug}`;
}

// ── Tracking ──────────────────────────────────────────────────────────────────

function loadTracking() {
  return existsSync(TRACKING) ? new Set(JSON.parse(readFileSync(TRACKING, "utf8"))) : new Set();
}

function saveTracking(set) {
  writeFileSync(TRACKING, JSON.stringify([...set].sort(), null, 2));
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  if (!TOKEN && !DRY_RUN) {
    console.error("❌ SANITY_API_TOKEN not found in .env.local");
    process.exit(1);
  }

  const allMd = readdirSync(ROOT)
    .filter(f => f.endsWith(".md") && f !== "SKILL.md")
    .sort();

  const published  = loadTracking();
  const candidates = allMd.filter(f => !published.has(f));

  if (!candidates.length) {
    console.log("✅ All articles already published — nothing to do.");
    return;
  }

  const toPublish = candidates.slice(0, COUNT);
  console.log(`📋 ${candidates.length} unpublished articles. Publishing ${toPublish.length}.\n`);

  if (!DRY_RUN) {
    console.log("🔧 Ensuring pillar and author exist...");
    await ensurePillar();
    await ensureAuthor();
  }

  const newlyPublished = [];

  for (const filename of toPublish) {
    console.log(`📄 ${filename}`);
    const article = parseArticle(resolve(ROOT, filename));
    if (!article) { console.log("  ⚠ Skipped.\n"); continue; }

    if (DRY_RUN) {
      console.log(`  [dry-run] article-${article.slug}`);
      console.log(`  Title:  ${article.title}`);
      console.log(`  Blocks: ${article.blocks.length}`);
      console.log(`  URL:    ${SITE_URL}/global-office/${article.slug}\n`);
      continue;
    }

    await ensureCountries(article.countries);
    const url = await publishArticle(article);
    published.add(filename);
    newlyPublished.push({ filename, title: article.title, url });
    console.log(`  ✓ ${article.title}`);
    console.log(`    → ${url}\n`);
  }

  if (!DRY_RUN && newlyPublished.length) {
    saveTracking(published);
    console.log(`✅ ${newlyPublished.length} article(s) published.`);
  } else if (DRY_RUN) {
    console.log(`✅ Dry run — ${toPublish.length} would be published.`);
  }

  const remaining = candidates.length - toPublish.length;
  if (remaining > 0) console.log(`📬 ${remaining} articles remain in queue.`);
}

main().catch(err => { console.error("❌", err.message); process.exit(1); });
